### DISCUSSION TITLE
Fried RAD after using w/ X-Pander 3?

### DISCUSSION THREAD

**Discussion by mstine** [frntc/RAD#8] [Q&A] [answered]
So, I've been happily using my RAD for a few weeks now, but I've missed having access to my Super Snapshot 5. My understanding is that the REU and SS5 operate in different address spaces, so they ought to behave together. So, I bought myself an X-Pander 3 from Retro Innovations, and it arrived today.

I plugged it in and seated the two cartridges, and then I flipped the dip switches to ON. Then I supplied power to the RPi as per usual, and switched on the C64. 

The SS5 booted as expected, and I hit F7 to escape to BASIC. I did not notice the typical REU message at the bottom of the screen, so I hit the menu button. Nothing. 

I thought maybe I was having a memory space conflict, so I powered down and swapped the IO1/2 jumpers that were going to one of the cartridges, thinking maybe it just wasn't getting addressed. Nope.

Tried the RAD without the X-Pander-3. Still not working. Now I'm worried.

Something told me to check the memory card. It wouldn't mount at all when I tried to attach it to my PC. I couldn't even repartition/format it that I could see. It seems fried. 

I made a new card. Still not working. But this card did survive, so whatever happened before didn't happen again. Granted, I'm no longer using the X-Pander-3 at this point.

Tried it with another PI (a 3B that I have that also happened to work before my 3A+ arrived). Nope.

If I insert the SD Card into the 3A+ and power it on WITHOUT the RAD attached, I see the disk access light flash a couple of times. With the RAD, it definitely does not flash.

I'm guessing the RAD is FUBAR. What I'm wondering is:

a) what the heck happened? Should I have not turned on the power switches on the X-Pander-3 for the REUs cartridge slot? Is that somehow different than plugging directly in to the C64? Either way, the power jumper was not connected, so I assumed the REU wouldn't be pulling power from the cartridge slot. Or am I missing something fundamental here?

b) is there a way to KNOW that the RAD is FUBAR? What should I test and how if so?

Thanks!

**Comment by frntc** [comment, marked as answer]
Am I right that you did not configure the RAD for REU-startup-mode? This might be the reason why there's a conflict with the SS5 during boot time (didn't I write "I recommend to NOT do this" ;-( ). Passive expanders (and the XP3 is one, right?) are always dangerous -- during Sidekick64 development I thrashed an EF3's CPLD... 

It's difficult to tell exactly what has happened. First you could check the HDMI output of the RPi and see if the firmware it's booting correctly. Next I could imagine that the LVC245 (which puts data on the bus) did not survive the fight against the SS5. Have you build the RAD yourself/are you able to desolder it?

**Reply by mstine** [reply]
You absolutely wrote that, and I absolutely read it. I got some false confidence from Robin's 8-Bit Show and Tell videos, where he uses the REU and SS5 in concert on his 128D. A few moments ago, it struck me that he was using an actual REU and an actual SS5. I use a RAD and a Kung Fu Flash with an SS5 ROM image. So these are 100% different circuits...

I did, in fact, configure for REU-startup mode. I've done that for quite some time.

So...when we're talking about danger here, my electronics knowledge is still too fuzzy to understand what you're saying. I thought address space conflicts were software issues and would lead to the computer acting up. It sounds like you're saying that they can lead to fried components. Or am I mixing things up?

Are you suggesting checking the HDMI output now? With or without the RAD attached? With or without the C64?

At any rate, I didn't build this guy myself. While learning to solder/desolder is on my long retrocomputing TODO list, I'm still in the breadboarding phase.

**Reply by frntc** [reply]
Let me give you an example of address space conflicts: say you use two simple cartridges which both bank in ROM at $8000 in parallel. When the CPU accesses memory at $8000 then both cartridges react and drive the data lines on the bus. So essentially two ICs (which want to output signals) work against each other.

I you have configured RAD to pure REU mode (and as long as you don't press the button) it's behaving relatively tame -- it only goes to DMA mode to check whether the C64 has finished initialization. It will not detect the "READY." prompt and thus do nothing (i.e. it's passive and not doing anything harmful). So thinking about this, I'd assume that a real SS5 would not cause damage in this configuration. I don't want to blame KFF for this, but I also don't know its internals nor how it behaves on the bus. Do you know if there are schematics for the X-Pander, or is really just plain passive?

And yes, I meant checking HDMI (1. without RAD, 2. with RAD, 3. with RAD + C64).

**Reply by mstine** [reply]
I think I see... so sometimes these conflicts end up being benign things like gibberish on the screen or a black screen, but sometimes they end up being surprising amounts of current hitting an IC because another one is also reacting, and there's not enough resistance in the circuit to account for that because, well, there shouldn't be another cartridge present. Is that more or less it?

So...you say, as long as you don't press the button. Well, the first thing I did when I didn't see it working was press the button. So, I'm hoping that didn't fry it.

Speaking of frying...both of these SD cards are now FUBAR. I'm unsure if it happened while in the Pi or while I was trying to diagnose them, but they will no longer hold a partition table. And I've tried a bunch of sysadmin-fu. So, I can't really do any additional testing until my new cards show up from our Amazon overlords tomorrow.

I assume you know the KFF schematics are out there, but just in case: https://github.com/KimJorgensen/KungFuFlash.

The X-Pander details are here: http://www.go4retro.com/products/x-pander-3/. They claim it is, in fact, a "passive cartridge port bus," and they also provide the details of all of the switches provided on each line. Maybe you can see something?

BTW - I think I figured out that Thomas Christoph made my board. I was searching for them on eBay based off of advice from a Discord, and I later put things together after finding this thread: https://www.lemon64.com/forum/viewtopic.php?t=81578. It doesn't seem like anyone is selling them right now, so I hope this sucker isn't dead. This probably shouldn't be the first project I attempt, lol.

**Reply by mstine** [reply]
So I did remember I have one more SD card on hand...in my KFF. 

I decided to put RPi OS on it and see if I can even get the Pis to boot normally. I can't. They're failing to mount the root FS. It looks like a common error, but thus far I've failed to fix it. Basically this:

`Kernel panic-not syncing: VFS: unable to mount root fs on unknown-block(179,6)`

**Reply by frntc** [reply]
The reason for ICs frying in this case is simply that two (or more) are outputting signals and it's out of specs if a "stronger" on forces the signals to something else than the other (current flows in the wrong direction). Afaik the KFF schematics are quite simple (no level shifters etc), but even when looking at the data sheets of the SoC I'm not sure if this would tell me much :-)

I fear if your RPi does not even boot an OS then this might be the problem (although the RPi is hidden completely behind a phalanx of ICs without directly connecting GPIOs to the C64).

When you're looking for a RAD: I maintain a "white list" of sellers (see readme) :-) For example restore-store.de currently has some, the owner is just on vacation this week or so (it's ~35USD including shipping to the US).

**Reply by mstine** [reply]
I really don't understand what's happening here...

So I decided not to futz around with RPi OS anymore and reformatted the card back to FAT. Replaced the RAD Firmware. I did the boot tests you suggested. FWIW, on the HDMI, all of them look the same:

```
logger: Circle 44.3 started on Raspberry Pi 3 Model A+ (AArch64)
00:00:00.28 timer: SpeedFactor is 3.57
00:00:00.80 emmc: Found a valid version 3.0x SD card
```

The confusion began when I flipped on the C64. The RAD menu appeared! The damn thing appears to be working. Did it just need a nap?

**Reply by mstine** [reply]
So I'm gonna call this question answered since everything seems to be fine. LOL.

At any rate, I did find what appears to be a configuration where the KFF and RAD can be attached to the X-Pander-3 and not kill each other.

First of all, the RAD seems to need to live in Port 1 to not suffer any signal degradation. Then, I flip on only:

GAME
EXROM
IO2
ROML
ROMH

The KFF is in Port 2 with:

POWER
POWER
GAME
EXROM
IO1
ROML
ROMH

So given that configuration, I am definitely not sending any extra power to the RAD, and only one IO line is switched on per device, so it seems like that should keep them segregated. That said, I need to study my manuals a bit more and make sure nothing else needs to be adjusted.

Thanks again for all of the help!

**Reply by frntc** [reply]
Great that it works (again). I didn't experience an RPi to be temporarily non-functional, but had this with a Teensy.

For the X-Pander: RAD should not need EXROM and ROML.

**Reply by mstine** [reply]
Confirmed! Thanks!
